### PR TITLE
feat: introduce `DynamicTimeoutLayer`

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Introduce `DynamicTimeoutLayer` for dynamic per-request timeout configuration via `timeout` header

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/servers/src/http/timeout.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(_) => Poll::Ready(Err(Elapsed::new().into())),
+        }
+    }
+}
+
+/// Applies a timeout to requests via the supplied inner service.
+#[derive(Debug, Clone)]
+pub struct DynamicTimeoutLayer {
+    timeout: Duration,
```
Comment (fengjiachun) [inline src/servers/src/http/timeout.rs:73]: I prefer `default_timeout`

File: src/servers/src/http/timeout.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+use tokio::time::Sleep;
+use tower::timeout::error::Elapsed;
+use tower::{BoxError, Layer, Service};
+
+/// [`Timeout`] response future
+///
+/// [`Timeout`]: crate::timeout::Timeout
+#[derive(Debug)]
+#[pin_project]
+pub struct ResponseFuture<T> {
```
Comment (fengjiachun) [inline src/servers/src/http/timeout.rs:33]: If we can't use `tower.timeout.future.ResponseFuture`, it's better to add a comment indicating the source of the code.

File: src/servers/src/http/timeout.rs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+impl<S> DynamicTimeout<S> {
+    /// Create a new [`DynamicTimeout`] with the given timeout
+    pub fn new(inner: S, timeout: Duration) -> Self {
+        DynamicTimeout { inner, timeout }
+    }
+}
+
+const USER_TIMEOUT_HEADER: &str = "timeout";
```
Comment (fengjiachun) [inline src/servers/src/http/timeout.rs:104]: Perhaps we can add the option to:  https://github.com/GreptimeTeam/greptimedb/blob/main/src/servers/src/http/header.rs#L25

File: src/servers/src/http/timeout.rs (hunk 2 above)
Comment (WenyXu) [reply src/servers/src/http/timeout.rs:37]: Added

File: src/servers/src/http/timeout.rs (hunk 1 above)
Comment (WenyXu) [reply src/servers/src/http/timeout.rs]: Solved in [3d3b66b](https://github.com/GreptimeTeam/greptimedb/pull/5006/commits/3d3b66b2c333010c86747a3119b1b1814f68ad88)

File: src/servers/src/http/timeout.rs (hunk 3 above)
Comment (WenyXu) [reply src/servers/src/http/timeout.rs]: Resolved in [3d3b66b](https://github.com/GreptimeTeam/greptimedb/pull/5006/commits/3d3b66b2c333010c86747a3119b1b1814f68ad88)

File: src/servers/src/http/header.rs [hunk 4]
```diff
@@ -39,6 +39,7 @@ pub mod constants {
 
     // LEGACY HEADERS - KEEP IT UNMODIFIED
     pub const GREPTIME_DB_HEADER_FORMAT: &str = "x-greptime-format";
+    pub const GREPTIME_DB_HEADER_TIMEOUT: &str = "x-greptime-timeout";
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/servers/src/http/header.rs:42]: It only allows disabling the entire timeout layer by setting options.timeout to zero.

```rust
        let timeout_layer = if self.options.timeout != Duration::default() {
            Some(ServiceBuilder::new().layer(DynamicTimeoutLayer::new(self.options.timeout)))
        } else {
            info!("HTTP server timeout is disabled");
            None
        };
```

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**